internal/graph/lookup.go

```diff
@@ -69,6 +73,115 @@ func (cl *ConcurrentLookup) Lookup(ctx context.Context, req ValidatedLookupReque
 	return resolved.Resp, resolved.Err
 }
 
+// SlowLookup performs a slow-path lookup request with the provided request and context.
+func (cl *ConcurrentLookup) SlowLookup(ctx context.Context, req ValidatedLookupRequest) (*v1.DispatchLookupResponse, error) {
+	if req.Subject.ObjectId == tuple.PublicWildcard {
+		resp := lookupResultError(req, NewErrInvalidArgument(errors.New("cannot perform lookup on wildcard")), emptyMetadata)
+		return resp.Resp, resp.Err
+	}
+
+	ds := datastoremw.MustFromContext(ctx)
+	it, err := ds.QueryTuples(ctx,
+		&v1_proto.RelationshipFilter{
+			ResourceType: req.ObjectRelation.Namespace,
+		},
+		req.Revision)
+	if err != nil {
+		resp := lookupResultError(req, err, emptyMetadata)
+		return resp.Resp, resp.Err
+	}
+	defer it.Close()
+
+	objects := tuple.NewONRSet()
+	for tpl := it.Next(); tpl != nil; tpl = it.Next() {
+		objects.Add(&v0.ObjectAndRelation{
+			Namespace: tpl.ObjectAndRelation.Namespace,
+			ObjectId:  tpl.ObjectAndRelation.ObjectId,
+			Relation:  req.ObjectRelation.Relation,
+		})
+	}
+
+	cancelCtx, checkCancel := context.WithCancel(ctx)
+	defer checkCancel()
+
+	g, checkCtx := errgroup.WithContext(cancelCtx)
+
+	type onrWithMeta struct {
+		onr  *v0.ObjectAndRelation
+		meta *v1.ResponseMeta
+	}
+	allowedONRs := make(chan onrWithMeta)
+
+	sem := make(chan struct{}, MaxConcurrentSlowLookupChecks)
```

Comment:
This seems like a good place to use https://pkg.go.dev/golang.org/x/sync/semaphore and not just a channel.